File: ros2/camera_control/camera_control/detector/circle_detector.py
import asyncio
import functools
import logging
from typing import List, Optional, Tuple

import cv2
import numpy as np


class CircleDetector:
    def __init__(self, logger: Optional[logging.Logger] = None):
        if logger:
            self._logger = logger
        else:
            self._logger = logging.getLogger(__name__)
            self._logger.setLevel(logging.INFO)

    async def detect(self, color_frame: np.ndarray) -> List[Tuple[int, int]]:
        """
        Detects centers of circles in a color image.

        Args:
            color_frame (np.ndarray): Color image.
        Returns:
            List[Tuple[int, int]]: List of detected center points.
        """
        return await asyncio.get_running_loop().run_in_executor(
            None,
            functools.partial(
                self._detect_sync,
                color_frame,
            ),
        )

    def _detect_sync(self, color_frame: np.ndarray) -> List[Tuple[int, int]]:
        # Convert to grayscale
        image = cv2.cvtColor(color_frame, cv2.COLOR_RGB2GRAY)

        # Apply a Gaussian blur to smooth the image and reduce noise
        blurred = cv2.GaussianBlur(image, (9, 9), 2)

        # Use HoughCircles to detect circles in the image
        circles = cv2.HoughCircles(
            blurred,
            cv2.HOUGH_GRADIENT,
            dp=1.2,  # inverse ratio of resolution (1 means same resolution)
            minDist=50,  # minimum distance between detected centers
            param1=100,  # higher threshold for Canny edge detector
            param2=30,  # accumulator threshold for circle detection
            minRadius=8,  # minimum circle radius
            maxRadius=50,  # maximum circle radius
        )

        circle_centers = []
        if circles is not None:
            circles = np.round(circles[0, :]).astype("int")
            circle_centers = [(x, y) for (x, y, r) in circles]

        return circle_centers

File: ros2/camera_control_cpp/src/detector/circle_detector.cpp
#include "camera_control_cpp/detector/circle_detector.hpp"

std::vector<std::tuple<int, int>> CircleDetector::detect(
    const cv::Mat& colorFrame) {
  // Convert to grayscale
  cv::Mat image;
  cv::cvtColor(colorFrame, image, cv::COLOR_BGR2GRAY);

  // Apply Gaussian blur to smooth the image and reduce noise
  cv::GaussianBlur(image, image, cv::Size(9, 9), 2);

  // Use HoughCircles to detect circles
  std::vector<cv::Vec3f> circles;
  cv::HoughCircles(
      image, circles, cv::HOUGH_GRADIENT,
      1.2,  // inverse ratio of resolution (1 means same resolution)
      50,   // minimum distance between detected centers
      100,  // higher threshold for Canny edge detector
      30,   // accumulator threshold for circle detection
      8,    // minimum circle radius
      50    // maximum circle radius
  );

  std::vector<std::tuple<int, int>> circleCenters;
  if (!circles.empty()) {
    for (const auto& circle : circles) {
      int x = cvRound(circle[0]);
      int y = cvRound(circle[1]);
      circleCenters.push_back(std::make_tuple(x, y));
    }
  }
  return circleCenters;
}
